File: typescript/libs/workspace/adapters/json/system/src/lib/workspace/Workspace.ts
import { Workspace as SystemWorkspace } from '@allors/workspace/adapters/system';
import { ISession, IWorkspaceServices } from '@allors/workspace/domain/system';
import { Database } from '../database/Database';
import { Session } from '../session/Session';

export class Workspace extends SystemWorkspace {

  constructor(database: Database, services: IWorkspaceServices) {
    super(database, services);
    
    this.services.onInit(this);
  }

  createSession(): ISession {
    return new Session(this, this.services.createSessionServices());
  }
}

File: dotnet/System/Workspace/Adapters/Allors.Workspace.Adapters.Local/Workspace/Workspace.cs
// <copyright file="LocalWorkspace.cs" company="Allors bv">
// Copyright (c) Allors bv. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Workspace.Adapters.Local
{
    using Ranges;

    public class Workspace : Adapters.Workspace
    {
        public Workspace(DatabaseConnection database, IWorkspaceServices services, IRanges<long> recordRanges) : base(database, services, recordRanges) => this.Services.OnInit(this);

        public new DatabaseConnection DatabaseConnection => (DatabaseConnection)base.DatabaseConnection;

        public long UserId => this.DatabaseConnection.UserId;

        public override ISession CreateSession() => new Session(this, this.Services.CreateSessionServices());
    }
}
